Fedora BRP script

gordonmessmer opened the pull request:
I've imported the Fedora BRP script using git am.  The result doesn't quite match the current Fedora version because there was one patch upstream that isn't in Fedora, but I think the result is correct.

torsava committed:
brp-python-bytecompile: Prepare for 2 digit minor versions (e.g. 3.10)

frenzymadness committed:
Remove support for the %_python_bytecompile_extra macro.

frenzymadness committed:
Use compileall from stdlib for Python >= 3.9

All enhancements from compileall2 are merged in Python 3.9.

kalev committed:
BRP Python Bytecompile: Avoid hardcoding /usr/bin prefix for python

Avoid using the full path and instead rely on PATH being correctly set
up to find the executable.

This fixes byte compilation for python2.7 when doing flatpak module
builds where python2.7 can be in either /usr/bin or /app/bin, depending
on how it's compiled.

baburton committed:
Adapt macros and BRP scripts for %topdir with spaces

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1947416

hroncok commented:
There are many very Fedora specific things in here. I wonder what @Conan-Kudo thinks about that. We don't want to make this repo Fedora-only.

Conan-Kudo commented on scripts/brp-python-bytecompile:
```diff
@@ -1,6 +1,15 @@
 errors_terminate=$2
+
+# Usage of %_python_bytecompile_extra is not allowed anymore
+# See: https://fedoraproject.org/wiki/Changes/No_more_automagic_Python_bytecompilation_phase_3
```
Remove this reference and explain properly in the comment.

Conan-Kudo commented on scripts/brp-python-bytecompile:
```diff
@@ -1,6 +1,15 @@
+# $default_python stays in the arguments for backward compatibility and $extra for the following check:
 extra=$3
+if [ 0$extra -eq 1 ]; then
+    echo -e "%_python_bytecompile_extra is discontinued, use %py_byte_compile instead.\nSee: https://fedoraproject.org/wiki/Changes/No_more_automagic_Python_bytecompilation_phase_3" >/dev/stderr
```
Remove the Change reference and explain properly here.

gordonmessmer replied:
Should we drop 8d2f0361ccd0f18f92e7bd71df673b53a229a724 entirely?

hroncok replied:
I'd keep it, but explain it in the error message maybe? Not sure.

Conan-Kudo commented on scripts/brp-python-bytecompile:
```diff
@@ -15,14 +24,60 @@ if [ -z "$depth" ] || [ "$depth" -le "1" ]; then
 fi
 
+# This function now implements Python byte-compilation in three different ways:
+# Python >= 3.4 and < 3.9 uses a new module compileall2 - https://github.com/fedora-python/compileall2
```
We need to somehow declare a dependency on `compileall2`.

hroncok replied:
We can drop the compileall2 thing entirely and call that a pro-active downstream only thing in Fedora, compileall2 has been merged to mainline comipleall. That would also solve the `/usr/lib/rpm/redhat/` thing.

Conan-Kudo replied:
That works for me.

gordonmessmer replied:
How much of 6d322d00 and 2c5fba02 should be kept?

hroncok replied:
I suggest we drop both of those commit entirely, but keep the Python 3.10+ branch.

Conan-Kudo commented on scripts/brp-python-bytecompile:
```diff
@@ -15,14 +24,60 @@ if [ -z "$depth" ] || [ "$depth" -le "1" ]; then
+		# -x and -e together implements the same functionality as the Filter class below
+		# -s strips $RPM_BUILD_ROOT from the path
+		# -p prepends the leading slash to the path to make it absolute
+		PYTHONPATH=/usr/lib/rpm/redhat/ $python_binary -B $options -m compileall2 -q -f $exclude -s $RPM_BUILD_ROOT -p / -e $RPM_BUILD_ROOT $python_libdir
```
We need to move this out of `/usr/lib/rpm/redhat`.

Conan-Kudo commented:
I don't like the commits for `/app` handling, that feels like the wrong solution for a cross-distro thing. Can we do this in a better way?

hroncok commented:
> I don't like the commits for `/app` handling, that feels like the wrong solution for a cross-distro thing. Can we do this in a better way?

We could pass `%_prefix` as an argument.

Conan-Kudo commented:
> > I don't like the commits for `/app` handling, that feels like the wrong solution for a cross-distro thing. Can we do this in a better way?
> 
> We could pass `%_prefix` as an argument.

That'd work for me.

gordonmessmer committed:
Allow an alternate %_prefix as an argument.

gordonmessmer commented:
Something like 27b458b0c744504b8ff402356e591cabb7dad789 ?  Or at this point should we implement option parsing rather than using positional parameters?

hroncok commented:
> Something like [27b458b](https://github.com/rpm-software-management/python-rpm-packaging/commit/27b458b0c744504b8ff402356e591cabb7dad789) ?

Yes, but also edit `platform.in` to pass it.

> Or at this point should we implement option parsing rather than using positional parameters?

I'd rather do that later, if at all.

gordonmessmer commented:
I've made another attempt at this...

gordonmessmer commented:
Checking in to see if there's any changes still needed for this PR.

hroncok commented:
I just didn't get to it yet, sorry :(

gordonmessmer commented:
No rush, of course.  Just wanted to make sure you weren't waiting on anything more from me.  :)

hroncok commented:
Finally, I got back to this. Sorry for the long delay. I'll have a look at the overall diff first, later at individual commits.

hroncok commented on scripts/brp-python-bytecompile:
```diff
@@ -15,14 +34,45 @@ if [ -z "$depth" ] || [ "$depth" -le "1" ]; then
 
+# This function now implements Python byte-compilation in two different ways:
+# Python < 3.9 (inc. Python 2) uses compileall module from stdlib with some hacks
+# When we drop support for Python 2, we'd be able to use all compileall2 features like:
```
```suggestion
# When we drop support for Python < 3.9, we'd be able to use all compileall features like:
```

hroncok commented on scripts/brp-python-bytecompile:
```diff
@@ -15,14 +34,45 @@ if [ -z "$depth" ] || [ "$depth" -le "1" ]; then
+# - -o 0 -o 1 to produce multiple files in one run - each with a different optimization level - instead of $options
+# - removed useless $depth - both compileall and compileall2 are limited by sys.getrecursionlimit()
+# These changes will make this script much simpler
+# In Python >= 3.9, compileall2 was merged back to standard library (compileall) so we can use it directly again.
```
```suggestion
```

hroncok reviewed (commented):
Since compileall2 usage was removed from here, I think we need to update the comments as well. Used suggestions to do so.

Other than that, the overall diff looks pretty good, thanks!

hroncok commented:
I'd personally squash *No more automagic Python bytecompilation (phase 3)* with *Document removal of support for the %_python_bytecompile_extra macro* to avoid Fedora specific stuff in git history. However, not a strong opinion.

hroncok commented:
Thanks for this. Overall feels almost ready to merge. I'll diff it with the Fedora script to verify all differences are "downstream specific".

Conan-Kudo commented:
> I'd personally squash _No more automagic Python bytecompilation (phase 3)_ with _Document removal of support for the %_python_bytecompile_extra macro_ to avoid Fedora specific stuff in git history. However, not a strong opinion.

I'd prefer this too, mostly because the comments and the output from the script winds up never showing Fedora specific stuff afterward in any commit.

encukou committed:
Use --hardlink-dupes in %py_byte_compile and brp-python-bytecompile

(for Python 3.9+)

Resolves: rhbz#1977895

encukou committed:
Use --hardlink-dupes for Python 3.4+ as well

encukou committed:
Remove unneeded arguments from the python_bytecompile function

- For depth, use sys.getrecursionlimit()
  - the default from bytecompile2
  - we'll get a RecursionError before it's exceeded, anyway
- Set real_libdir locally

hroncok committed:
Add a note: Python 3.11+ no longer needs PYTHONHASHSEED=0

Implemented in:  https://github.com/python/cpython/pull/27926

We keep using it thou, because this is Python version agnostic.

Once we drop support for anything older than 3.11, we can remove it.
That'll be around ~2030. Assuming the world still exists by then.

hroncok committed:
Bytecompilation: Pass --invalidation-mode=timestamp to compileall

(Only on Python 3.7+, where it exists and matters.)

This will replace patch 328 in Python and fix https://bugzilla.redhat.com/2133850

hroncok committed:
Add a script to clamp source mtimes, invoke it from the bytecompilation BRP/macro

https://fedoraproject.org/wiki/Changes/ReproducibleBuildsClampMtimes

hroncok committed:
Bytecompilation: Compile Python files in parallel, according to %_smp_mflags

Conan-Kudo commented:
@gordonmessmer This seems to fail CI after rebasing... Can you take a look at this?

gordonmessmer committed:
Remove reference to modular repos in github workflow.

gordonmessmer committed:
Download the current setuptools sdist in pythondistdeps test.

gordonmessmer committed:
Use "main" for tox workflows.

gordonmessmer committed:
Update fsleyes version for pythondistdeps test.

gordonmessmer committed:
Remove python 3.8, because it can't download setuptools 75.6.

gordonmessmer committed:
Remove LegacyVersion requirements from pyreq2rpm test data.

gordonmessmer committed:
Update stderr for corrupted.dist-info in pythondistdeps tests.

gordonmessmer committed:
Add newer Python releases to tox_env GitHub workflow.

gordonmessmer commented:
Tests are running again, but I think there's a branch protection rule expecting a run of tox_test with py38.

Is there a reason to preserve py38?  I'd have to modify one of the changes I made to restore the tests.

hroncok commented:
Hey @gordonmessmer. I am so sorry for ignoring this for so long. In fact, the revival caught me by surprise.

I wonder if the repository makes sense after all. 

Do openSUSE folks use it? We certainly currently don't. And I feel bad about it, but there are always more urgent things :(

-----

> Is there a reason to preserve py38? 

No.

Conan-Kudo commented:
> Do openSUSE folks use it? We certainly currently don't. And I feel bad about it, but there are always more urgent things :(

I would like distributions (Fedora, openSUSE, Mageia, etc.) all to use this.

gordonmessmer committed:
Upgrade github actions/cache to v4, as v2 has been deprecated.

gordonmessmer commented:
> I wonder if the repository makes sense after all.

I don't know, but I volunteered to resync the two and now my branch is *mostly* in sync with Fedora's script, again. 

>>    Is there a reason to preserve py38?

> No.

Then I think a repository owner needs to remove the branch protection rule that requires the py38 test.

gordonmessmer commented:
> I don't like the commits for `/app` handling, that feels like the wrong solution for a cross-distro thing. Can we do this in a better way?

... and I imported the app handling again, having forgotten why the prefix spec was done.  I'll back that out shortly.

gordonmessmer committed:
Make logical operators consistent in brp-python-bytecompile.

gordonmessmer commented:
Let me know if there's any fedora-specific stuff that should be backed out.

Conan-Kudo commented on scripts/brp-python-bytecompile:
```diff
@@ -1,34 +1,110 @@
+        # -x and -e together implements the same functionality as the Filter class below
+        # -s strips $RPM_BUILD_ROOT from the path
+        # -p prepends the leading slash to the path to make it absolute
+        PYTHONPATH=/usr/lib/rpm/redhat/ $python_binary -B -m $compileall_module $compileall_flags -o 0 -o 1 -q -f $exclude -s "$RPM_BUILD_ROOT" -p / --hardlink-dupes $invalidation_option -e "$RPM_BUILD_ROOT" "$python_libdir"
```
We can't assume `/usr/lib/rpm/redhat`

gordonmessmer replied:
How about referring to a vendor-neutral location (from platform.in this time)?

gordonmessmer replied:
hm... actually, I probably need to also pull the compileall2 module from python-rpm-macros and possibly add it to b3063f3f or just after that commit.

gordonmessmer committed:
Expect the mtime clamp module in /usr/lib/python-packaging

gordonmessmer committed:
Minor style fixes suggested by ShellCheck. Mostly, these consist
of preferring '[[' to '[' in bash scripts. Other changes include
quoting unquoted variables, and explicitly specifying bash as the
interpreter for scripts that use features not defined in POSIX sh

Fixes SC2046, SC3001, and SC2292

gordonmessmer committed:
The "exclude" variable in python_bytecompile is no longer used, so remove it.

This resolves ShellCheck SC2089 and SC2090 warnings.

ncoghlan commented on scripts/brp-python-bytecompile:
```diff
@@ -53,64 +123,38 @@ EOF
 
 shopt -s nullglob
-for python_libdir in `find "$RPM_BUILD_ROOT" -type d|grep -E "/usr/lib(64)?/python[0-9]\.[0-9]$"`;
+find "$RPM_BUILD_ROOT" -type d -print0|grep -z -E "${prefix}/lib(64)?/python[0-9]\.[0-9]+$" | while read -d "" python_libdir;
```
The missing `+` in the Python version regex is a particularly notable absence for anything involving Python 3.10 or later (which now means all Python versions that are still supported by the upstream developers)
